VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+---
+vip: 10
+title: Storage node registration (on-chain)
+author: Chris Were <chris@verida.io>
+discussions-to: #dev-protocol on Verida Discord
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-03-22
+---
+
+# Overview
+
+The selection of storage nodes occurs when a DID is first created and when a context is connected to for the first time. The storage nodes are used to store the actual DID Documents and the same nodes are typically (but not necessarily) used to store data for each application context.
+
+This VIP provides a specification for storing the available storage nodes on-chain.
+
+This is phase 2 of [VIP-7: Storage node discovery and selection (off-chain)](./vip-7.md)
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
+
+This method registers a new endpoint on the network:
+
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
+6. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
```

Comment:
I think, `unregisteredDatetime` is not necessary because we can get time in the blockchain itself.